tabpy-server/tabpy_server/app/util.py

```diff
@@ -35,3 +37,58 @@ def validate_cert(cert_file_path):
     if now > not_after:
         log_and_raise(https_error +
                       'The certificate provided expired on {}.'.format(not_after), RuntimeError)
+
+
+def parse_pwd_file(pwd_file_name):
+    '''
+    Parses passwords file and returns set of credentials.
+
+    Parameters
+    ----------
+    pwd_file_name : str
+        Passwords file name.
+
+    Returns
+    -------
+    succeeded : bool
+        True if specified file was parsed successfully.
+        False if there were any issues with parsing specified file.
+
+    credentials : dict
+        Credentials from the file. Empty if succeeded is False.
+    '''
+    logger.info('Parsing passwords file {}...'.format(pwd_file_name))
+
+    if not os.path.isfile(pwd_file_name):
+        logger.fatal('Passwords file {} not found'.format(pwd_file_name))
+        return False, {}
+
+    credentials = {}
+    with open(pwd_file_name) as pwd_file:
+        pwd_file_reader = csv.reader(pwd_file, delimiter=' ')
+        for row in pwd_file_reader:
+            # skip empty lines
+            if len(row) == 0:
+                continue
+
+            # skip commented lines
+            if row[0][0] == '#':
+                continue
+
+            if len(row) != 2:
+                logger.error(
+                    'Incorrect entry "{}" '
+                    'in password file'.format(row))
+                return False, {}
+
+            login = row[0].lower()
+            if login in credentials:
+                logger.error(
+                    'Multiple entries for username {} '
+                    'in password file'.format(login))
+                return False, {}
+
+            credentials[login] = row[1]
+            logger.debug('Found username {}'.format(login))
+
+    return True, credentials
```

Comment:
So back to previous conversation around security, I think the README updates you made are great! But the problem is that regardless of OS permissions a user could technically still deploy a function to import and execute this function to return back the credentials of all users so long as they can make a reasonable guess at the file name (which probably wouldn't be that hard).

I think it would really help security a lot if credentials were never returned here nor exposable through anything that exists in the global namespace of any module

Reply:
Yeah, it's an architectural problem caused by TabPy executing user scripts in the same context where the Web server runs. 

It is not that bad as we are dealing with hashes of passwords and TabPy never knows actual passwords. Another limiting factor for an attacker is he/she needs to know at least one pair of login/password to be able to evaluate a script. And since all the users are the same to TabPy as it is implemented now (we don't have any authorization or resources restrictions) knowing additional login/password won't give an attacker any benefit.

Reply:
Yea for sure. So my point is that it's just pretty easy for someone to get a hold of these hashed passwords, and that makes things difficult but obviously not impossible to target a particular user (say CEO of company) and try to match the hash with their own computing resources.

It would be a much better design if there was a database locked with a password available only to the deployed server which stored passwords and would be checked against per user on demand rather than having them persist in the memory space of Python

Reply:
Agreed, but it also could be configuration nightmare if we require people to configure a DB and access to it considering TabPy setup itself can be challenging for some platforms.

Again, even with using basic access authentication here we see it more like auditing mechanism at the moment.

Reply:
Yea makes sense. SQLite would be pretty easy to leverage here for future reference

Agreed it's an improvement over what's in place regardless